redbot/core/commands/commands.py

```diff
@@ -485,15 +485,22 @@ async def prepare(self, ctx):
         if not await self.can_run(ctx, change_permission_state=True):
             raise CheckFailure(f"The check functions for command {self.qualified_name} failed.")
 
-        if self.cooldown_after_parsing:
-            await self._parse_arguments(ctx)
-            self._prepare_cooldowns(ctx)
-        else:
-            self._prepare_cooldowns(ctx)
-            await self._parse_arguments(ctx)
         if self._max_concurrency is not None:
             await self._max_concurrency.acquire(ctx)
-        await self.call_before_hooks(ctx)
+
+        try:
+            if self.cooldown_after_parsing:
+                await self._parse_arguments(ctx)
+                self._prepare_cooldowns(ctx)
+            else:
+                self._prepare_cooldowns(ctx)
+                await self._parse_arguments(ctx)
+
+            await self.call_before_hooks(ctx)
+        except:
+            if self._max_concurrency is not None:
+                await self._max_concurrency.release(ctx)
+            raise
```

Comment:
```suggestion
        except (
            discord.ClientException,
            discord.ext.commands.UserInputError,
            discord.ext.commands.CommandOnCooldown,
        ) as e:
            if self._max_concurrency is not None:
                await self._max_concurrency.release(ctx)
            raise e
```
Raw except should instead specify the exceptions it's expecting.
I included the three that `_parse_arguments` and `_prepare_cooldowns` could throw, but all might not be appropriate.

Also, please modify the imports to import `UserInputError` and `CommandOnCooldown` rather than accepting this suggestion outright.

Reply:
This has been brought up before and it will remain as is because this is pretty much a copy of what d.py is doing and this way it's easier to maintain and ensures that the behavior between d.py and Red remains identical.